src/jira/client/axios.ts

```diff
@@ -1,74 +1,60 @@
 import Logger from "bunyan";
-import axios, { AxiosError, AxiosInstance } from "axios";
+import axios, { AxiosError, AxiosInstance, AxiosRequestConfig } from "axios";
 
 import url from "url";
 import statsd from "../../config/statsd";
 import { getLogger } from "../../config/logger";
 import { metricHttpRequest } from "../../config/metric-names";
 import { createQueryStringHash, encodeSymmetric } from "atlassian-jwt";
-import {urlParamsMiddleware} from "../../util/axios/common-middleware";
+import { urlParamsMiddleware } from "../../util/axios/url-params";
 
 const instance = process.env.INSTANCE_NAME;
 const iss = `com.github.integration${instance ? `.${instance}` : ""}`;
 
-// TODO: type hack to fix custom implementation of URL templating vars
-declare module "axios" {
-	interface AxiosRequestConfig {
-		urlParams?: Record<string, string>;
-	}
-}
-
 /**
  * Middleware to create a custom JWT for a request.
  *
  * @param {string} secret - The key to use to sign the JWT
  */
-function getAuthMiddleware(secret: string) {
-	return (
-		/**
-		 * @param {import("axios").AxiosRequestConfig} config - The config for the outgoing request.
-		 * @returns {import("axios").AxiosRequestConfig} Updated axios config with authentication token.
-		 */
-		(config) => {
-			const { query, pathname } = url.parse(config.url, true);
-
-			const jwtToken = encodeSymmetric(
-				{
-					...getExpirationInSeconds(),
-					iss,
-					qsh: createQueryStringHash({
-						method: config.method,
-						pathname: pathname || undefined,
-						query
-					})
-				},
-				secret
-			);
-
-			return {
-				...config,
-				headers: {
-					...config.headers,
-					Authorization: `JWT ${jwtToken}`
-				}
-			};
-		}
-	);
-}
+const getAuthMiddleware = (secret: string, instance: AxiosInstance) =>
+	(config: AxiosRequestConfig): AxiosRequestConfig => {
+		// Generate full URI based on current config
+		const uri = instance.getUri(config);
```

Comment:
@KAbakumovAtl This is specifically where the problem lied.  It was using `config.url` to parse the path and querystring, but that's an optional parameter and most would use the `params` option instead for queries instead of adding it inline.  By doing this, the path/query could be wrong depending on the situation.

Now we can actually get the full URI that will be generated by Axios before the request and parse that instead since there is an encoding step that happens for all params.

As for tests, there's probably a few we can add around the axios client since we'd need to test it in conjunction with the other middleware which can interfere.  I'll add some in.

Reply:
That sounds like it could affect Push API as well. Is it because we didn't use query parameters and the url didnt' change?

Reply:
yep.